### PR TITLE
fix: remove type coercion of case expression in Expr::Schema

### LINKED ISSUE #5821: Remove type coercion of case expression in Expr::Schema
As part of fixing https://github.com/apache/arrow-datafusion/issues/5733 in https://github.com/apache/arrow-datafusion/pull/5734 we installed a workaround that manipulates the calculation of the type of a Case expression

A better solution would be to fix the expression initially (by coercing the types correctly), but that was blocked on https://github.com/apache/arrow-datafusion/issues/5733

Once https://github.com/apache/arrow-datafusion/issues/5733 is complete, we should remove the workaround put in Expr::schema

Details below:

              The type coercion is implemented as an optimizer - and it changes schemas of particular `Expr`essions.
And it **takes care** of `Expr::Case`.

But `Expr::Projection` caches the schema in itself - so it is not updated! Therefore we get a mismatch between declared schema vs batch schema.

Essentially this PR makes `Expr::Case` to **not** change a schema during the coercion optimizer.

_Originally posted by @mslapek in https://github.com/apache/arrow-datafusion/pull/5734#discussion_r1153540699_

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #5821.

# Rationale for this change

relate with #6613 

# What changes are included in this PR?

- fix create project with new schema during analysis
- remove type coercion of case expression in Expr::Schema

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/scalar.slt [hunk 1]
```diff
@@ -646,27 +646,27 @@ SELECT CASE WHEN NULL THEN 'foo' ELSE 'bar' END
 bar
 
 # case_expr_with_null()
-query I
+query ?
```
Comment (jackwener) [inline datafusion/core/tests/sqllogictests/test_files/scalar.slt:649]: It's strange. I can't understand it.

I exec sql to confirm `type coercion` is right

```sql
 select arrow_typeof(case when b is null then null else b end) from (select a,b from (values (1,null),(2,3)) as t (a,b)) a;

+------------------------------------------------------------+
| arrow_typeof(CASE WHEN a.b IS NULL THEN NULL ELSE a.b END) |
+------------------------------------------------------------+
| Int64                                                      |
| Int64                                                      |
+------------------------------------------------------------+
```

File: datafusion/core/tests/sql/mod.rs [hunk 2]
```diff
@@ -971,13 +971,8 @@ async fn plan_and_collect(ctx: &SessionContext, sql: &str) -> Result<Vec<RecordB
 async fn execute_to_batches(ctx: &SessionContext, sql: &str) -> Vec<RecordBatch> {
     let df = ctx.sql(sql).await.unwrap();
 
-    // We are not really interested in the direct output of optimized_logical_plan
-    // since the physical plan construction already optimizes the given logical plan
-    // and we want to avoid double-optimization as a consequence. So we just construct
-    // it here to make sure that it doesn't fail at this step and get the optimized
-    // schema (to assert later that the logical and optimized schemas are the same).
-    let optimized = df.clone().into_optimized_plan().unwrap();
-    assert_eq!(df.logical_plan().schema(), optimized.schema());
```
Comment (jackwener) [inline datafusion/core/tests/sql/mod.rs:980]: we shouldn't check schema here.
Because analyzer will change schema

**Comment by jackwener** [comment]
cc @alamb @liukun4515 @mslapek

File: datafusion/core/tests/sql/mod.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/tests/sql/mod.rs:980]: I agree this check predated the existence of the analyzer

**Review by alamb — APPROVED** [review]
Makes sense to me -- thank you @jackwener

**Pull request merged**